### PR TITLE
Mixer: Add assertions regarding num_samples versus buffer length

### PR DESCRIPTION
**Pull request by Pokechu22**
Compare #10664. I haven't been able to reproduce any buffer overflows (other than the audio dumping one from #10741), but hopefully these assertions will make them more obvious when they happen.

(I didn't add any assertions to `Mixer::MixerFifo::PushSamples` because that code *should* already be free of overflows due to the masking.)

### CODE REVIEW

File: Source/Core/AudioCommon/Mixer.cpp [hunk 1]
```diff
@@ -282,8 +292,10 @@ void Mixer::PushStreamingSamples(const short* samples, unsigned int num_samples)
 void Mixer::PushWiimoteSpeakerSamples(const short* samples, unsigned int num_samples,
                                       unsigned int sample_rate_divisor)
 {
-  short samples_stereo[MAX_SAMPLES * 2];
+  static std::array<short, MAX_SAMPLES * 2> samples_stereo;
```
Comment (AdmiralCurtiss) [inline Source/Core/AudioCommon/Mixer.cpp:295]: `static`?

Comment (Pokechu22) [reply Source/Core/AudioCommon/Mixer.cpp:295]: We don't really need to allocate it on the stack, but I guess there's no harm in doing so either. I don't remember intentionally changing it to `static` for what it's worth.

Also, on further unrelated investigation, it's not possible for there to be more than 40 wiimote samples at a time (due to the maximum size of the output report combined with 4-bit ADPCM). So this array is probably way larger than it really needs to be.

Comment (AdmiralCurtiss) [reply Source/Core/AudioCommon/Mixer.cpp:295]: I'd rather you leave this on the stack, especially when it's much smaller anyway (maybe make that change and add an assert?), we have enough hard-to-find global state as-is...

File: Source/Core/AudioCommon/Mixer.cpp [hunk 2]
```diff
@@ -282,9 +292,14 @@ void Mixer::PushStreamingSamples(const short* samples, unsigned int num_samples)
...
                                       unsigned int sample_rate_divisor)
 {
-  short samples_stereo[MAX_SAMPLES * 2];
-
-  if (num_samples < MAX_SAMPLES)
+  // Max 20 bytes/speaker report, may be 4-bit ADPCM so multiply by 2
+  static constexpr u32 MAX_SPEAKER_SAMPLES = 20 * 2;
+  std::array<short, MAX_SPEAKER_SAMPLES * 2> samples_stereo;
+
+  ASSERT_MSG(AUDIO, num_samples < MAX_SPEAKER_SAMPLES,
```
Comment (AdmiralCurtiss) [inline Source/Core/AudioCommon/Mixer.cpp:299]: This needs to be `num_samples <= MAX_SPEAKER_SAMPLES`... Actually I think this applies to all of the asserts as well as the `if` a few lines down.

**Comment by AdmiralCurtiss** [comment]
Seems good, let's see if anything or anyone triggers these.

**Pull request merged**